File: src/machine_list.cpp
#include "machine_list.h"

void MachineryList::Advance(Terrain * level, TankList *)
{
    /* Advance everything */
    this->items.ForEach([level](auto & item) { item.Advance(level); });
}

void MachineryList::Draw(Surface * surface)
{ 
    this->items.ForEach([surface](auto & item) { item.Draw(surface); });
}

Machine * MachineryList::GetMachineAtPoint(Position position) 
{
    Machine * result = nullptr;
    this->items.ForEach([position, &result](auto & harvester) {
        if (harvester.TestCollide(position))
            result = &harvester;
    });
    return result;
}

MachineTemplate * MachineryList::GetMachineTemplateAtPoint(Position position)
{
    MachineTemplate * result = nullptr;
    this->items.ForEachConvertibleTo<MachineTemplate>([position, &result](auto & harvester) {
        if (harvester.TestCollide(position))
            result = &harvester;
    });
    return result;
}

File: dotnet/src/TunnelTanks.Core/Entities/Machines/MachineList.cs
namespace TunnelTanks.Core.Entities.Machines;

using TunnelTanks.Core.Types;
using TunnelTanks.Core.Terrain;

public class MachineList
{
    private readonly List<Machine> _machines = new();

    public IReadOnlyList<Machine> Machines => _machines;

    public Machine Add(Machine machine)
    {
        _machines.Add(machine);
        return machine;
    }

    public void Advance(Terrain terrain, TimeSpan dt)
    {
        for (int i = _machines.Count - 1; i >= 0; i--)
        {
            if (!_machines[i].IsAlive) { _machines.RemoveAt(i); continue; }
            _machines[i].Advance(terrain, dt);
        }
    }

    public Machine? TestCollide(Position pos)
    {
        foreach (var m in _machines)
            if (m.IsAlive && m.TestCollide(pos))
                return m;
        return null;
    }

    public void Draw(uint[] surface, int surfaceWidth, int surfaceHeight)
    {
        foreach (var m in _machines)
            m.Draw(surface, surfaceWidth, surfaceHeight);
    }

    public void RemoveAll() => _machines.Clear();
}
